### PR TITLE
Knowledge Distillation bug fixes

### PR DESCRIPTION
**Pull request by SKholkin**
Changes:

- Fixes failure of running Knowledge Distillation in DataParallel mode. More precisely fixing tensor device placing logic.
- Fixes failure of Knowledge Distillation tests for CPU only case.
- Changing logic of storage Knowledge Distillation loss values. Now compression_ctrl.loss() may be called many times per one iteration.

### CODE REVIEW

**Comment by ljaljushkin** [comment]
new tests are missing. please add

File: nncf/torch/knowledge_distillation/knowledge_distillation_loss.py [hunk 1]
```diff
@@ -103,12 +103,11 @@ def forward(self) -> torch.Tensor:
         """
         loss = self._kd_loss_handler.get_kd_loss()
         if len(loss) == 0:
-            raise RuntimeError('Empty list of loss tensors for KDLoss. Most likely compression_ctrl.loss()'
-                               ' was called while model was in eval mode')
+            warnings.warn("Empty list of loss tensors for Knowledge Distillation Loss.", RuntimeWarning)
```
Comment (ljaljushkin) [inline nncf/torch/knowledge_distillation/knowledge_distillation_loss.py:106]:
```suggestion
            warnings.warn("Knowledge Distillation Loss is not calculated.", RuntimeWarning)
```

File: nncf/torch/knowledge_distillation/knowledge_distillation_loss.py [hunk 2]
```diff
@@ -80,10 +79,11 @@ def match_fn(obj):
         orig_model_loss_outputs = list(map(lambda x: x.getter(), filter(
             match_fn, orig_model_outputs_nested_obj_indexing.get_flat_nested_obj_indexing())))
         if len(orig_model_loss_outputs) == 0 or len(compressed_model_loss_outputs) == 0:
-            return torch.zeros([], device=device)
+            return torch.zeros([], device='cpu')
```
Comment (ljaljushkin) [inline nncf/torch/knowledge_distillation/knowledge_distillation_loss.py:82]: why cpu?

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/torch/knowledge_distillation/knowledge_distillation_loss.py (hunk 1 above)
Comment (SKholkin) [reply nncf/torch/knowledge_distillation/knowledge_distillation_loss.py]: Done

File: nncf/torch/knowledge_distillation/knowledge_distillation_loss.py (hunk 2 above)
Comment (SKholkin) [reply nncf/torch/knowledge_distillation/knowledge_distillation_loss.py]: This code block was deleted at all

**Comment by SKholkin** [comment]
> new tests are missing. please add

i've added special test for device related logic.

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please retry a build

File: tests/torch/test_knowledge_distillation.py [hunk 3]
```diff
@@ -269,3 +278,71 @@ def test_kd_sparsity_statistics(algo: str):
            getattr(statistics_with_kd, algo).model_statistics.sparsity_level
     assert getattr(statistics, algo).model_statistics.sparsity_level_for_layers ==\
            getattr(statistics_with_kd, algo).model_statistics.sparsity_level_for_layers
+
+
+@pytest.mark.parametrize("device_placing", ['before', 'after'])
+@pytest.mark.parametrize("inference_type", ['cpu', 'single_GPU', 'DP', 'DDP'])
+def test_model_device_before_create_compressed_model(device_placing, inference_type):
+    if not torch.cuda.is_available() and not inference_type == 'cpu':
+        return
```
Comment (vshampor) [inline tests/torch/test_knowledge_distillation.py:287]: Use `pytest.skip` instead and give a reason why the test is skipped, like it is being done in other tests.

**Review by vshampor — CHANGES_REQUESTED** [review]

**Review by vshampor — APPROVED** [review]
